File: Backend/app/schemas/gate_entry.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from decimal import Decimal


class GateEntryMaterialBase(BaseModel):
    material_code: Optional[str] = None
    material_name: str
    quantity: Decimal = Field(..., gt=0)
    uom: str


class GateEntryMaterialCreate(GateEntryMaterialBase):
    pass


class GateEntryMaterialResponse(GateEntryMaterialBase):
    id: str
    gate_entry_id: str
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class GateEntryBase(BaseModel):
    entry_type: str = Field(..., description="material, courier, visitor, jobwork_return, subcontract_return, delivery, machine_spare")
    vendor: str = Field(..., min_length=1, max_length=255)
    vehicle_no: Optional[str] = Field(None, max_length=50)
    driver_name: Optional[str] = Field(None, max_length=255)
    linked_document: Optional[str] = Field(None, max_length=100, description="PO/Invoice/DC/AWB number")
    destination_department: str = Field(..., description="Store, QA, Maintenance, Production, Admin")
    remarks: Optional[str] = None


class GateEntryCreate(GateEntryBase):
    materials: List[GateEntryMaterialCreate] = Field(..., min_items=1)


class GateEntryUpdate(BaseModel):
    entry_type: Optional[str] = None
    vendor: Optional[str] = None
    vehicle_no: Optional[str] = None
    driver_name: Optional[str] = None
    linked_document: Optional[str] = None
    destination_department: Optional[str] = None
    status: Optional[str] = None
    remarks: Optional[str] = None
    materials: Optional[List[GateEntryMaterialCreate]] = None


class GateEntryResponse(GateEntryBase):
    id: str
    entry_number: str
    status: str
    materials: List[GateEntryMaterialResponse] = []
    created_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class GateEntryListItem(BaseModel):
    """Simplified gate entry for list view"""
    id: str
    entry_number: str
    entry_type: str
    vendor: str
    vehicle_no: Optional[str]
    driver_name: Optional[str]
    destination_department: str
    status: str
    linked_document: Optional[str]
    material_count: int
    first_material_name: Optional[str]
    total_items: Optional[Decimal]
    created_at: datetime
    
    class Config:
        from_attributes = True


class GateEntryStatusUpdate(BaseModel):
    status: str = Field(..., description="arrived, under_verification, accepted, rejected")
    remarks: Optional[str] = None


class GateEntryStats(BaseModel):
    """Gate entry statistics"""
    total_entries: int
    arrived: int
    under_verification: int
    accepted: int
    rejected: int
    today_entries: int
    by_type: dict
    by_department: dict

File: Frontend/src/lib/api/gateEntry.ts
import { apiClient } from './client';

export interface GateEntryMaterial {
    id?: string;
    gate_entry_id?: string;
    material_code?: string;
    material_name: string;
    quantity: number;
    uom: string;
    created_at?: string;
    updated_at?: string;
}

export interface GateEntryResponse {
    id: string;
    entry_number: string;
    entry_type: string;
    vendor: string;
    vehicle_no?: string;
    driver_name?: string;
    linked_document?: string;
    destination_department: string;
    status: string;
    remarks?: string;
    photos?: string[];
    materials: GateEntryMaterial[];
    created_by?: string;
    created_at: string;
    updated_at: string;
}

export interface GateEntryListItem {
    id: string;
    entry_number: string;
    entry_type: string;
    vendor: string;
    vehicle_no?: string;
    driver_name?: string;
    destination_department: string;
    status: string;
    linked_document?: string;
    material_count: number;
    first_material_name?: string;
    total_items?: number;
    created_at: string;
}

export interface GateEntryStats {
    total_entries: number;
    arrived: number;
    under_verification: number;
    accepted: number;
    rejected: number;
    today_entries: number;
    by_type: Record<string, number>;
    by_department: Record<string, number>;
}

export interface CreateGateEntryData {
    entry_type: string;
    vendor: string;
    vehicle_no?: string;
    driver_name?: string;
    linked_document?: string;
    destination_department: string;
    remarks?: string;
    photos?: string[];
    materials: {
        material_code?: string;
        material_name: string;
        quantity: number;
        uom: string;
    }[];
}

export interface UpdateGateEntryData {
    entry_type?: string;
    vendor?: string;
    vehicle_no?: string;
    driver_name?: string;
    linked_document?: string;
    destination_department?: string;
    status?: string;
    remarks?: string;
    photos?: string[];
    materials?: {
        material_code?: string;
        material_name: string;
        quantity: number;
        uom: string;
    }[];
}

export const gateEntryApi = {
    // List all gate entries
    list: async (params?: {
        page?: number;
        limit?: number;
        status?: string;
        entry_type?: string;
        search?: string;
        date_from?: string;
        date_to?: string;
    }): Promise<GateEntryListItem[]> => {
        const queryParams = new URLSearchParams();
        if (params) {
            Object.entries(params).forEach(([key, value]) => {
                if (value !== undefined) queryParams.append(key, value.toString());
            });
        }
        return apiClient.get<GateEntryListItem[]>(`/gate-entries/?${queryParams.toString()}`);
    },

    // Get gate entry statistics
    getStats: async (): Promise<GateEntryStats> => {
        return apiClient.get<GateEntryStats>('/gate-entries/stats');
    },

    // Get single gate entry
    get: async (id: string): Promise<GateEntryResponse> => {
        return apiClient.get<GateEntryResponse>(`/gate-entries/${id}`);
    },

    // Create new gate entry
    create: async (data: CreateGateEntryData): Promise<GateEntryResponse> => {
        return apiClient.post<GateEntryResponse>('/gate-entries/', data);
    },

    // Update gate entry
    update: async (id: string, data: UpdateGateEntryData): Promise<GateEntryResponse> => {
        return apiClient.put<GateEntryResponse>(`/gate-entries/${id}`, data);
    },

    // Update gate entry status
    updateStatus: async (id: string, data: {
        status: string;
        remarks?: string;
    }): Promise<GateEntryResponse> => {
        return apiClient.put<GateEntryResponse>(`/gate-entries/${id}/status`, data);
    },

    // Delete gate entry
    delete: async (id: string): Promise<{ message: string }> => {
        return apiClient.delete<{ message: string }>(`/gate-entries/${id}`);
    }
};
